File: src/classes/backoffs.ts
import { BackoffOptions, MinimalJob } from '../interfaces';
import { BackoffStrategy } from '../types';

export interface BuiltInStrategies {
  [index: string]: (delay: number, percentage?: number) => BackoffStrategy;
}

export class Backoffs {
  static builtinStrategies: BuiltInStrategies = {
    fixed: function (delay: number) {
      return function (): number {
        return delay;
      };
    },

    exponential: function (delay: number) {
      return function (attemptsMade: number): number {
        return Math.round(Math.pow(2, attemptsMade - 1) * delay);
      };
    },

    jitter: function (delay: number, percentage = 1) {
      return function (attemptsMade: number): number {
        const maxDelay = Math.round(Math.pow(2, attemptsMade - 1) * delay);
        const minDelay = maxDelay * (1 - percentage);

        return Math.floor(Math.random() * maxDelay * percentage + minDelay);
      };
    },
  };

  static normalize(
    backoff: number | BackoffOptions,
  ): BackoffOptions | undefined {
    if (Number.isFinite(<number>backoff)) {
      return {
        type: 'fixed',
        delay: <number>backoff,
      };
    } else if (backoff) {
      return <BackoffOptions>backoff;
    }
  }

  static calculate(
    backoff: BackoffOptions,
    attemptsMade: number,
    err: Error,
    job: MinimalJob,
    customStrategy?: BackoffStrategy,
  ): Promise<number> | number | undefined {
    if (backoff) {
      const strategy = lookupStrategy(backoff, customStrategy);

      return strategy(attemptsMade, backoff.type, err, job);
    }
  }
}

function lookupStrategy(
  backoff: BackoffOptions,
  customStrategy?: BackoffStrategy,
): BackoffStrategy {
  if (backoff.type in Backoffs.builtinStrategies) {
    return Backoffs.builtinStrategies[backoff.type](
      backoff.delay!,
      backoff.percentage,
    );
  } else if (customStrategy) {
    return customStrategy;
  } else {
    throw new Error(
      `Unknown backoff strategy ${backoff.type}.
      If a custom backoff strategy is used, specify it when the queue is created.`,
    );
  }
}

File: python/bullmq/backoffs.py
from typing import Any, Callable
from bullmq.types import BackoffOptions

import math


class Backoffs:

    builtin_strategies = {
        "fixed": lambda delay: lambda attempts_made, type, err, job: delay,
        "exponential": lambda delay: lambda attempts_made, type, err, job: int(round(pow(2, attempts_made - 1) * delay))
    }

    @staticmethod
    def normalize(backoff: int | BackoffOptions) -> BackoffOptions | None:
        if type(backoff) == int and math.isfinite(backoff):
            return {
                "type": "fixed",
                "delay": backoff
            }
        elif backoff:
            return backoff

    @staticmethod
    async def calculate(
        backoff: BackoffOptions | None,
        attempts_made: int,
        err: Any,
        job: Any,
        customStrategy: Callable[..., int] | None,
    ) -> int | None:
        if backoff:
            strategy = lookup_strategy(backoff, customStrategy)
            return strategy(attempts_made, backoff.get("type"), err, job)


def lookup_strategy(
    backoff: BackoffOptions,
    custom_strategy: Callable[..., int] | None,
) -> Callable[..., int]:
    backoff_type = backoff.get("type")
    if backoff_type in Backoffs.builtin_strategies:
        return Backoffs.builtin_strategies[backoff_type](backoff.get("delay"))
    elif custom_strategy:
        return custom_strategy
    else:
        raise Exception(f"Unknown backoff strategy {backoff_type}. " +
                        "If a custom backoff strategy is used, specify it when the queue is created.")
